File: internal/ui/navigation.go
package ui

// Navigation stack management methods for the Model.
// These methods provide push/pop navigation and breadcrumb generation.

// screenName returns a human-readable name for a screen.
func screenName(s Screen) string {
	switch s {
	case ScreenMainMenu:
		return "Main Menu"
	case ScreenGameTypeSelect:
		return "New Game"
	case ScreenBotSelect:
		return "Bot Difficulty"
	case ScreenColorSelect:
		return "Choose Color"
	case ScreenFENInput:
		return "Load Game"
	case ScreenGamePlay:
		return "Game"
	case ScreenGameOver:
		return "Game Over"
	case ScreenSettings:
		return "Settings"
	case ScreenSavePrompt:
		return "Save Game"
	case ScreenResumePrompt:
		return "Resume Game"
	case ScreenDrawPrompt:
		return "Draw Offer"
	case ScreenBvBBotSelect:
		return "Bot vs Bot Setup"
	case ScreenBvBGameMode:
		return "Game Mode"
	case ScreenBvBGridConfig:
		return "Grid Layout"
	case ScreenBvBGamePlay:
		return "Bot vs Bot"
	case ScreenBvBStats:
		return "Statistics"
	case ScreenBvBViewModeSelect:
		return "View Mode"
	case ScreenBvBConcurrencySelect:
		return "Concurrency Select"
	default:
		return "Unknown"
	}
}

// pushScreen navigates to a new screen, saving the current screen to the navigation stack.
// Use this for forward navigation where the user should be able to go back.
func (m *Model) pushScreen(newScreen Screen) {
	// Don't push if we're already on the same screen
	if m.screen == newScreen {
		return
	}
	// Add current screen to stack before navigating
	m.navStack = append(m.navStack, m.screen)
	m.screen = newScreen
}

// popScreen returns to the previous screen in the navigation stack.
// If the stack is empty, it returns to the main menu.
// Returns the screen that was navigated to.
func (m *Model) popScreen() Screen {
	if len(m.navStack) == 0 {
		m.screen = ScreenMainMenu
		return ScreenMainMenu
	}

	// Pop the last screen from the stack
	lastIndex := len(m.navStack) - 1
	previousScreen := m.navStack[lastIndex]
	m.navStack = m.navStack[:lastIndex]
	m.screen = previousScreen
	return previousScreen
}

// clearNavStack clears the navigation stack.
// Use this when starting a new game or returning to a clean state.
func (m *Model) clearNavStack() {
	m.navStack = nil
}

// breadcrumb generates a breadcrumb string showing the navigation path.
// Returns an empty string if at the main menu or if the stack is empty.
func (m Model) breadcrumb() string {
	if m.screen == ScreenMainMenu || len(m.navStack) == 0 {
		return ""
	}

	// Build breadcrumb from the last item in the stack (direct parent)
	// We show just the parent > current for simplicity
	if len(m.navStack) > 0 {
		parent := m.navStack[len(m.navStack)-1]
		return screenName(parent) + " > " + screenName(m.screen)
	}

	return screenName(m.screen)
}

// canGoBack returns true if there's a previous screen to navigate back to.
func (m Model) canGoBack() bool {
	return len(m.navStack) > 0
}

File: rust/crates/ui/src/navigation.rs
//! Navigation stack management (port of Go `navigation.go`).

use crate::app::{build_main_menu_options, App};
use crate::state::Screen;

/// Human-readable name for a screen (Go `screenName`).
pub fn screen_name(s: Screen) -> &'static str {
    match s {
        Screen::MainMenu => "Main Menu",
        Screen::GameTypeSelect => "New Game",
        Screen::BotSelect => "Bot Difficulty",
        Screen::ColorSelect => "Choose Color",
        Screen::FenInput => "Load Game",
        Screen::GamePlay => "Game",
        Screen::GameOver => "Game Over",
        Screen::Settings => "Settings",
        Screen::SavePrompt => "Save Game",
        Screen::DrawPrompt => "Draw Offer",
        Screen::BvBBotSelect => "Bot vs Bot Setup",
        Screen::BvBGameMode => "Game Mode",
        Screen::BvBGridConfig => "Grid Layout",
        Screen::BvBGamePlay => "Bot vs Bot",
        Screen::BvBStats => "Statistics",
        Screen::BvBViewModeSelect => "View Mode",
        Screen::BvBConcurrencySelect => "Concurrency Select",
    }
}

impl App {
    /// Navigates forward, pushing the current screen (Go `pushScreen`).
    pub fn push_screen(&mut self, new_screen: Screen) {
        if self.screen == new_screen {
            return;
        }
        self.nav_stack.push(self.screen);
        self.screen = new_screen;
    }

    /// Returns to the previous screen (Go `popScreen`).
    pub fn pop_screen(&mut self) -> Screen {
        match self.nav_stack.pop() {
            Some(prev) => {
                self.screen = prev;
                self.restore_menu_state();
                prev
            }
            None => {
                self.screen = Screen::MainMenu;
                self.restore_menu_state();
                Screen::MainMenu
            }
        }
    }

    /// Restores menu options/selection for the current screen (Go `restoreMenuState`).
    pub fn restore_menu_state(&mut self) {
        match self.screen {
            Screen::GameTypeSelect => {
                self.menu_options = vec![
                    "Player vs Player".into(),
                    "Player vs Bot".into(),
                    "Bot vs Bot".into(),
                ];
            }
            Screen::BvBBotSelect => {
                self.menu_options = vec!["Easy".into(), "Medium".into(), "Hard".into()];
            }
            Screen::BvBGameMode => {
                self.menu_options = vec!["Single Game".into(), "Multi-Game".into()];
            }
            Screen::BvBGridConfig => {
                self.menu_options = vec![
                    "1x1".into(),
                    "2x2".into(),
                    "2x3".into(),
                    "2x4".into(),
                    "Custom".into(),
                ];
            }
            Screen::BvBConcurrencySelect => {}
            Screen::BvBViewModeSelect => {
                self.menu_options = vec![
                    "Grid View".into(),
                    "Single Board".into(),
                    "Stats Only".into(),
                ];
            }
            Screen::MainMenu => {
                self.menu_options = build_main_menu_options();
            }
            Screen::BotSelect => {
                self.menu_options = vec!["Easy".into(), "Medium".into(), "Hard".into()];
            }
            Screen::ColorSelect => {
                self.menu_options = vec!["Play as White".into(), "Play as Black".into()];
            }
            Screen::Settings => {
                self.menu_options = vec![format!("Theme: {}", self.theme.name)];
            }
            _ => {}
        }
        self.menu_selection = 0;
        self.error_msg.clear();
    }

    /// Clears the navigation stack (Go `clearNavStack`).
    pub fn clear_nav_stack(&mut self) {
        self.nav_stack.clear();
    }

    /// Generates a breadcrumb string (Go `breadcrumb`).
    pub fn breadcrumb(&self) -> String {
        if self.screen == Screen::MainMenu || self.nav_stack.is_empty() {
            return String::new();
        }
        let parent = *self.nav_stack.last().unwrap();
        format!("{} > {}", screen_name(parent), screen_name(self.screen))
    }

    /// Whether there is a screen to go back to (Go `canGoBack`).
    pub fn can_go_back(&self) -> bool {
        !self.nav_stack.is_empty()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::app::App;
    use std::sync::mpsc::channel;

    fn app() -> App {
        let (tx, _rx) = channel();
        App::new(config::Config::default(), tx)
    }

    #[test]
    fn screen_names() {
        assert_eq!(screen_name(Screen::MainMenu), "Main Menu");
        assert_eq!(screen_name(Screen::GameTypeSelect), "New Game");
        assert_eq!(screen_name(Screen::BvBGamePlay), "Bot vs Bot");
    }

    #[test]
    fn push_and_pop() {
        let mut a = app();
        assert_eq!(a.screen, Screen::MainMenu);
        assert!(!a.can_go_back());
        a.push_screen(Screen::GameTypeSelect);
        assert_eq!(a.screen, Screen::GameTypeSelect);
        assert!(a.can_go_back());
        a.push_screen(Screen::BotSelect);
        assert_eq!(a.screen, Screen::BotSelect);
        let prev = a.pop_screen();
        assert_eq!(prev, Screen::GameTypeSelect);
        assert_eq!(a.screen, Screen::GameTypeSelect);
    }

    #[test]
    fn push_same_screen_is_noop() {
        let mut a = app();
        a.push_screen(Screen::MainMenu);
        assert!(a.nav_stack.is_empty());
    }

    #[test]
    fn pop_empty_returns_main_menu() {
        let mut a = app();
        a.screen = Screen::Settings;
        let s = a.pop_screen();
        assert_eq!(s, Screen::MainMenu);
        assert_eq!(a.screen, Screen::MainMenu);
    }

    #[test]
    fn breadcrumb_string() {
        let mut a = app();
        assert_eq!(a.breadcrumb(), "");
        a.push_screen(Screen::GameTypeSelect);
        assert_eq!(a.breadcrumb(), "Main Menu > New Game");
    }

    #[test]
    fn clear_nav_stack_works() {
        let mut a = app();
        a.push_screen(Screen::GameTypeSelect);
        a.push_screen(Screen::BotSelect);
        a.clear_nav_stack();
        assert!(!a.can_go_back());
    }
}
